kafka-client/src/main/java/dev/responsive/kafka/api/async/internals/contexts/StreamThreadProcessorContext.java

```diff
@@ -0,0 +1,118 @@
+/*
+ * Copyright 2024 Responsive Computing, Inc.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package dev.responsive.kafka.api.async.internals.contexts;
+
+import dev.responsive.kafka.api.async.internals.events.DelayedForward;
+import dev.responsive.kafka.api.async.internals.events.DelayedWrite;
+import dev.responsive.kafka.api.async.internals.stores.AsyncKeyValueStore;
+import java.util.HashMap;
+import java.util.Map;
+import org.apache.kafka.streams.processor.StateStore;
+import org.apache.kafka.streams.processor.api.ProcessorContext;
+import org.apache.kafka.streams.processor.internals.InternalProcessorContext;
+import org.apache.kafka.streams.processor.internals.ProcessorNode;
+import org.apache.kafka.streams.processor.internals.ProcessorRecordContext;
+import org.apache.kafka.streams.state.KeyValueStore;
+
+/**
+ * A wrapper around the original processor context to be used by the StreamThread.
+ * This context handles everything related to initialization of the processor
+ * (eg {@link #getStateStore(String)}) as well as the "finalization" of async events,
+ * ie preparing for (and executing) delayed forwards or writes.
+ * Delayed forwards are intercepted by the other kind of async context, which is
+ * unique to a given AsyncThread, called the {@link AsyncThreadProcessorContext}
+ * <p>
+ * Threading notes:
+ * -For use by StreamThreads only
+ * -One per physical AsyncProcessor
+ *  (ie one per async processor per partition per StreamThread)
+ */
+public class StreamThreadProcessorContext<KOut, VOut>
+    extends DelegatingInternalProcessorContext<KOut, VOut> {
+
+  private final Map<String, AsyncKeyValueStore<?, ?>> storeNameToAsyncStore = new HashMap<>();
+  private final ProcessorNode<?, ?, ?, ?> asyncProcessorNode;
+
+  public StreamThreadProcessorContext(
+      final ProcessorContext<KOut, VOut> delegate
+  ) {
+    super((InternalProcessorContext<KOut, VOut>) delegate);
+
+    asyncProcessorNode = super.currentNode();
+  }
+
+  @Override
+  @SuppressWarnings("unchecked")
+  public <S extends StateStore> S getStateStore(final String name) {
+    final S userDelegate = super.getStateStore(name);
```

Comment:
this needs to check if `storeNameToAsyncStore` contains this entry already and use the created one if it does

Reply:
oof, good catch. fixed